### ISSUE TITLE
Coloring issue on 1.8 Implementation

### ISSUE THREAD

**Issue by DevDrizzy** [MegavexNetwork/scoreboard-library#66]
I'm currently using your packetevents-implementation with 1.8 paper. It displays perfectly fine to legacy clients. However, it just breaks for the modern clients for some reason. It has a weird line length as well. (Changing to the 1.8 native implementation fixes this, however then that implementation has coloring issues sometimes.)

**1.8 Client**
![image](https://github.com/user-attachments/assets/2faf2569-e203-42cf-9ce3-31c4612b5aa0)

**1.20.4 Client**
Here it cut off the party members size along with weird spaces.
![image](https://github.com/user-attachments/assets/a8c22a4d-4e78-4a92-808f-32f819924054)

**1.8 Native implementation Color issue**
Here, it should be ``&fOnline: &c0`` and ``&fLeader: &cName``, but it's white. This didn't happen in the packetevents-implementation.
![yBHjt55](https://github.com/user-attachments/assets/97038222-ce8c-457f-9e0a-21eef78c89b5)

**Implementation**
```java
package xyz.refinedev.api.scoreboard.component;

import com.google.common.base.Preconditions;

import lombok.Getter;

import net.kyori.adventure.text.Component;

import net.megavex.scoreboardlibrary.api.sidebar.Sidebar;
import net.megavex.scoreboardlibrary.api.sidebar.component.ComponentSidebarLayout;
import net.megavex.scoreboardlibrary.api.sidebar.component.SidebarComponent;
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.SidebarAnimation;

import org.bukkit.entity.Player;

import xyz.refinedev.api.scoreboard.adapter.ScoreboardAdapter;
import xyz.refinedev.api.scoreboard.animation.ScoreboardAnimation;
import xyz.refinedev.api.scoreboard.utils.AnimationUtil;
import xyz.refinedev.api.scoreboard.utils.ColorUtil;

import java.util.*;

@Getter
public class DefaultScoreboardComponent {

    private final Map<String, ScoreboardAnimation> animations = new HashMap<>();

    private final Sidebar sidebar;
    private final Player player;
    private final ScoreboardAdapter adapter;

    private boolean animated;
    private ComponentSidebarLayout componentSidebar;
    private SidebarComponent title;
    private SidebarAnimation<Component> titleAnimation;

    /**
     * Create a scoreboard for the given player.
     *
     * @param adapter {@link ScoreboardAdapter} The adapter to use for this scoreboard.
     * @param sidebar {@link Sidebar} The sidebar to display the scoreboard on.
     * @param player  {@link Player} The player to display the scoreboard to.
     */
    public DefaultScoreboardComponent(ScoreboardAdapter adapter, Sidebar sidebar, Player player) {
        this.sidebar = sidebar;
        this.adapter = adapter;
        this.player = player;

        this.title = SidebarComponent.staticLine(this.adapter.getTitle(this.player)));
    }

    public void reload() {
        this.sidebar.removePlayer(this.player);
        this.setup();
        this.sidebar.addPlayer(this.player);
    }

    // Called every tick
    public void tick() {
        if (this.sidebar.closed()) return;

        List<String> lines = this.adapter.getLines(this.player);
        if (lines.isEmpty()) {
            if (this.sidebar.players().contains(this.player)) {
                this.sidebar.removePlayer(this.player);
            }
            return;
        } else {
            if (!this.sidebar.players().contains(this.player)) {
                this.sidebar.addPlayer(this.player);
            }
        }

        SidebarComponent component = createComponent(lines);
        this.componentSidebar = new ComponentSidebarLayout(title, component);

        // Update sidebar title & lines
        this.componentSidebar.apply(this.sidebar);
    }

    private SidebarComponent createComponent(List<String> lines) {
        SidebarComponent.Builder builder = SidebarComponent.builder();
        for (String line : lines) {
            if (line.isEmpty() || line.equals(" ")) {
                builder.addBlankLine();
                continue;
            }

            if (line.contains("<") || line.contains(">") || line.contains("%")) {
                builder.addDynamicLine(() -> ColorUtil.translate(line));
            } else {
                builder.addStaticLine(ColorUtil.translate(line));
            }
        }
        return builder.build();
    }
}
```

Please note that there are no animated lines being used.

**Comment by vytskalt** [comment]
What client version is the last screenshot taken on?

**Comment by vytskalt** [comment]
https://github.com/ViaVersion/ViaVersion/issues/4179

**Comment by DevDrizzy** [comment]
> What client version is the last screenshot taken on?

If you mean the color one, 1.8.9 optifine

**Comment by vytskalt** [comment]
Can you send me the exact line strings so I can try to reproduce it myself?

**Comment by DevDrizzy** [comment]
> Can you send me the exact line strings so I can try to reproduce it myself?

Mine is based on config, and some internal placeholders. But here,

```java
    public static List<String> LOBBY_SCOREBOARD = Arrays.asList(
            "&7&m----------------------",
            "&c&lLobby",
            " &fOnline: &c{online}",
            " &fQueuing: &c{in_queues}",
            " &fFighting: &c{in_fights}",
            "",
            " &fCoins: &c{coins} &e⛀",
            " &fTokens: &c{event_tokens}",
            "",
            " &fLevel: {division_color}{level}✮ &f({division_color}{next_percentage}%&f)",
            " {division_progress}",
            "",
            "&7&otest.refinedev.xyz",
            "&7&m----------------------"
    );
    
    // Use legacy serialize to convert to component for the API.
    private List<String> getLobbyLines() {
        List<String> entries = new ArrayList<>();
        for ( String line : ScoreboardConfig.LOBBY_SCOREBOARD ) {
               line = line.replace("{online}", Bukkit.getOnlinePlayers().size())
                              .replace("{in_queues}", Integer.toString(0));
                              .replace("{in_fights}", Integer.toString(0));
                entries.add(line);
            }
        }
        return entries;
    }
    ```

**Comment by DevDrizzy** [comment]
I have changed the formatting code in [LegacyLocaleLine](https://github.com/MegavexNetwork/scoreboard-library/blob/main/implementation/src/main/java/net/megavex/scoreboardlibrary/implementation/sidebar/line/locale/LegacyLocaleLine.java#L65) to the following, which has resolved the coloring issues. It also works perfectly fine with modern clients.

```java
    @Override
    public void value(@NotNull Component renderedComponent) {
        // I really need to rewrite this
        String line = legacySection().serialize(renderedComponent);
        String prevPlayer = player;

        int maxLength = 16;

        if (line.isEmpty()) {
            this.prefix = COLOR_CODES[info.line()] + ChatColor.RESET;
            this.suffix = "";
        } else if (line.length() <= maxLength) {
            this.prefix = line + StringUtils.repeat(" ", maxLength - line.length());
            this.suffix = "";

            if (this.currentValue != null && this.currentValue.length() > 32) {
                this.player = info.player();
            }
        } else {
            // Prevent splitting color codes
            boolean endsWithSection = line.charAt(15) == LegacyComponentSerializer.SECTION_CHAR;

            int prefixEnd = endsWithSection ? 15 : 16;
            this.prefix = line.substring(0, prefixEnd);

            String suffixTmp = line.substring(prefixEnd);
            ChatColor chatColor = null;

            if (suffixTmp.length() >= 2 && suffixTmp.charAt(0) == LegacyComponentSerializer.SECTION_CHAR) {
                chatColor = ChatColor.getByChar(suffixTmp.charAt(1));
            }

            String color = ChatColor.getLastColors(this.prefix);
            boolean addColor = chatColor == null || chatColor.isFormat();

            suffix = (addColor ? (color.isEmpty() ? ChatColor.RESET.toString() : color) : "") + suffixTmp;
        }

        // Something went wrong, just cut to prevent client crash/kick
        if (prefix.length() > maxLength) {
            prefix = prefix.substring(0, maxLength);
        }
        if (suffix.length() > maxLength) {
            suffix = suffix.substring(0, maxLength);
        }

        currentValue = prefix + suffix;

        if (!player.equals(prevPlayer)) {
            oldPlayer = prevPlayer;
        }
    }
```

I have also replaced the ``limitLegacyText`` method usage in [TeamsPacketAdapterImpl.java](https://github.com/MegavexNetwork/scoreboard-library/blob/main/versions/v1_8_R3/src/main/java/net/megavex/scoreboardlibrary/implementation/packetAdapter/v1_8_R3/TeamsPacketAdapterImpl.java#L81) to the above code just in case.

**Comment by vytskalt** [comment]
Those changes significantly lower the max line length, which is something I'd like to avoid.

Before:
![image](https://github.com/user-attachments/assets/70b7496f-fc61-40a2-86c2-fbc15acb9c68)

After:
![image](https://github.com/user-attachments/assets/df584595-e21a-4585-b8f8-2a46d862c21a)

**Comment by DevDrizzy** [comment]
> Those changes significantly lower the max line length, which is something I'd like to avoid.
> 
> Before: ![image](https://private-user-images.githubusercontent.com/35194709/376025204-70b7496f-fc61-40a2-86c2-fbc15acb9c68.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4MTQ3MDYsIm5iZiI6MTcyODgxNDQwNiwicGF0aCI6Ii8zNTE5NDcwOS8zNzYwMjUyMDQtNzBiNzQ5NmYtZmM2MS00MGEyLTg2YzItZmJjMTVhY2I5YzY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDEzVDEwMTMyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhNDE1ZjI0NjM4MGM0ZDgxN2E1ZDc3YzE0ODNhNzI0MjQ1NzU2MjA4MmNjZDRjOGNlYjJhYzgxOTYwNGEwM2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hplQXokeVj93EuRmnlopjLUtjqfkuQH9EFaiAwB2fjA)
> 
> After: ![image](https://private-user-images.githubusercontent.com/35194709/376025127-df584595-e21a-4585-b8f8-2a46d862c21a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4MTQ3MDYsIm5iZiI6MTcyODgxNDQwNiwicGF0aCI6Ii8zNTE5NDcwOS8zNzYwMjUxMjctZGY1ODQ1OTUtZTIxYS00NTg1LWI4ZjgtMmE0NmQ4NjJjMjFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDEzVDEwMTMyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzOWEzYzgzNzM4N2RlYWI2NjA0NjMzMmU5ZWE3NzQyNjE3NjUyMTYyMzI1MDNkNjQ2ODAyNzI5YWY2ZDIyNjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iLNEF67RD5I2GnMX-JJB0LX6QHneAf5UnQcM-7NXSs8)

for 1.8? or modern clients?

**Comment by vytskalt** [comment]
for 1.12.2 and below

**Comment by DevDrizzy** [comment]
I'll try and add back the name I suppose, I'll let you know how that goes

**Comment by DevDrizzy** [comment]
Coloring issue has been resolved by the commit [50d567a](https://github.com/MegavexNetwork/scoreboard-library/commit/50d567a4afa6ab63dfcec0a9cc54dd4a0613dccc).
